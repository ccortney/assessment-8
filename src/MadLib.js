import MadLibForm from "./MadLibForm";
import MadLibStory from "./MadLibStory";
import MadLibSelector from "./MadLibSelector";
import { useState } from "react";

/* Main component. MadLib Component starts by rendering buttons to choose mad lib story.
* When a story is chosen and the story information is saved in state, that information
* is sent to MadLibForm Component and the form is shown to users. When the form is 
* submitted, the user input is sent to MadLibStory Component and the story is constructed
* and displayed. */
const MadLib = () => {
    const [storyInfo, setStoryInfo] = useState();
    const [storyUserData, setStoryUserData] = useState();
    const [showStoryOptions, setShowStoryOptions] = useState(true);
    const [showForm, setShowForm] = useState(false);
    const [showStory, setShowStory] = useState(false);
    
    const chooseStory = (story) => {
        setStoryInfo(story);
        setShowForm(true);
        setShowStoryOptions(false);
    }

    const createStory = (formData) => {
        setStoryUserData(formData)
        setShowForm(false)
        setShowStory(true)
    }

    const clearStory = () => {
        setStoryUserData();
        setShowStory(false)
        setShowStoryOptions(true);
    }

    return (
        <div>
            <h1>Madlibs!</h1>
            {showStoryOptions ? (<MadLibSelector chooseStory = {chooseStory}/>) : null}
            {showForm ? (<MadLibForm createStory = {createStory} inputs={storyInfo.values}/>) : null}
            {showStory ? (<MadLibStory story={storyInfo.template} data={storyUserData} clearStory={clearStory} />) : null}

        </div>

    )
}

export default MadLib;
