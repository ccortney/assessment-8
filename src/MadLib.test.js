import { render, fireEvent } from '@testing-library/react';
import MadLib from './MadLib';

// smoke test
test('renders without crashing', () => {
  render(<MadLib />);
});

// snapshop test
test ('matches snapshot', () => {
  const {asFragment} = render(<MadLib/>);
  expect(asFragment()).toMatchSnapshot();
})

// more tests
test ('initial rendering of page', () => {
    const madLib = render(<MadLib/>);

    // story selector buttons should be on page
    expect(madLib.queryByText("Short but Sweet")).toBeInTheDocument();

    // form and story should not be on page
    expect(madLib.queryByText("Create Story")).not.toBeInTheDocument();
    expect(madLib.queryByText("Reset")).not.toBeInTheDocument();
})

test ('if form renders when story is selected', () => {
    const madLib = render(<MadLib/>);

    // story is selected
    const storyButton = madLib.queryByText("Short but Sweet");
    fireEvent.click(storyButton)

    // buttons should not be rendered
    expect(madLib.queryByText("Short but Sweet")).not.toBeInTheDocument();

    // the form should be rendered
    expect(madLib.queryByText("Create Story")).toBeInTheDocument();
})

test ('if story renders when form is submitted', () => {
    const madLib = render(<MadLib/>);

    // story is selected and form is rendered
    const storyButton = madLib.queryByText("Short but Sweet");
    fireEvent.click(storyButton)

    // form is filled out
    const nounInput = madLib.getByPlaceholderText("singular noun");
    fireEvent.change(nounInput, {target: {value: "pencil"}})
    const adjectiveInput = madLib.getByPlaceholderText("adjective");
    fireEvent.change(adjectiveInput, {target: {value: "fuzzy"}})
    const verbInput = madLib.getByPlaceholderText("plural verb");
    fireEvent.change(verbInput, {target: {value: "sings"}})

    // form is submitted
    const submitButton = madLib.queryByText("Create Story");
    fireEvent.click(submitButton)

    // story is rendered
    expect(madLib.queryByText("The fuzzy pencil sings!")).toBeInTheDocument();

    // form should no longer be displayed
    expect(madLib.queryByText("Create Story")).not.toBeInTheDocument();
})

test ('if reset button resets the page', () => {
    const madLib = render(<MadLib/>);

    // story is selected and form is rendered
    const storyButton = madLib.queryByText("Short but Sweet");
    fireEvent.click(storyButton)

    // form is filled out
    const nounInput = madLib.getByPlaceholderText("singular noun");
    fireEvent.change(nounInput, {target: {value: "pencil"}})
    const adjectiveInput = madLib.getByPlaceholderText("adjective");
    fireEvent.change(adjectiveInput, {target: {value: "fuzzy"}})
    const verbInput = madLib.getByPlaceholderText("plural verb");
    fireEvent.change(verbInput, {target: {value: "sings"}})

    // form is submitted
    const submitButton = madLib.queryByText("Create Story");
    fireEvent.click(submitButton)

    // story is rendered
    expect(madLib.queryByText("The fuzzy pencil sings!")).toBeInTheDocument();

    // reset button is clicked
    const resetButton = madLib.queryByText("Reset");
    fireEvent.click(resetButton)

    // story should not be rendered
    expect(madLib.queryByText("The fuzzy pencil sings!")).not.toBeInTheDocument();

    // story selector buttons should be rendered
    expect(madLib.queryByText("Short but Sweet")).toBeInTheDocument();
})