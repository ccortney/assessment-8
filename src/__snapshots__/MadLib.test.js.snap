// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Madlibs!
    </h1>
    <div>
      <button>
        A Love Story
      </button>
      <button>
        Animal Escapades
      </button>
      <button>
        Short but Sweet
      </button>
    </div>
  </div>
</DocumentFragment>
`;
